### ISSUE TITLE
Alternator: DescribeTable can return spurious RANGE key for GSI

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#5320]
In the current implementation, when `DescribeTable`  wants to list the table's key columns (`KeySchema`) we don't have the originally requested schema explicitly saved anywhere, so ` describe_key_schema()` attempts to recreate this information from the Scylla table's schema.

But there is one problem with this when describing the `KeySchema` of a GSI (note that `DescribeTable` returns the schema of the base table and of all its indexes). The problem is in one specific case: In DynamoDB, one can have a base table with hash key "p", and create a GSI with a *different* hash key "x" *and no range key.* In Scylla, we had to, internally, also add "p" as a range key to the materialized view - because a view key must include the base's key columns. This means that with the current implementation, this extra range key "x" is returned by `DescribeTable` - yet it shouldn't have been. We have an xfailing test case - `test_gsi_2_describe_table_schema` - to reproduce this case.

**Comment by nyh** [comment]
@bartpeeters commented in https://github.com/scylladb/scylladb/issues/11455#issuecomment-1237810598:

It is one of the things that makes Alternator not work with Terraform, after creating the table using Terraform and then applying the plan again it will pickup 'changes' to the index:

```
- global_secondary_index {
    - hash_key           = "ownerId" -> null
    - non_key_attributes = [] -> null
    - range_key          = "channelId" -> null
    - read_capacity      = 0 -> null
    - write_capacity     = 0 -> null
  }
+ global_secondary_index {
    + hash_key           = "ownerId"
    + name               = "ownerId-index"
    + non_key_attributes = []
    + projection_type    = "ALL"
  }
```

Terraform index resource code:

```
global_secondary_index {
  name            = "ownerId-index"
  hash_key        = "ownerId"
  projection_type = "ALL"
}
```

But I don't think this is the only thing tripping up Terraform, you also don't see the name or projection_type properties on the current table.

**Comment by fee-mendes** [comment]
A user tripped into this today - further adding that tools **do** expect an Index `KeySchema` to contain only the attributes specified upon creation:

> We have a process that checks to see if secondary indexes declared in our metadata have been applied in the backing datastore, and since the schema differs, our startup check believes that the target index has not been created.

**Comment by swasik** [comment]
Assigning to @ScyllaPiotr to find the owner. Somebody was already working on fixes to DescribeTable, @radoslawcybulski if I remember correctly. Let us discuss it today during the team sync.

**Comment by nyh** [comment]
@swasik asked me to do this issue.

**Comment by nyh** [comment]
> [@swasik](https://github.com/swasik) asked me to do this issue.

Not any more ;-)

**Comment by nyh** [comment]
For reference, here is a paragraph I wrote in a private email an hour ago when I was planning to work on this issue:

_"This is a fairly trivial problem, and we already have a test for it. But I need to think how to solve this in the least "ugly" way:
The problem is that it's not a simple bug in DescribeTable but rather information we don't have (whether the range key is a "real" one requested by the user) - as you may know, Alternator doesn't keep separate information about its tables, and all the information we have are the CQL schema we generate for the table - and also tags. I can definitely store the extra information using hidden tags (we've been doing this more and more - like recently Amnon added tags to store ProvisionedThroughput in commit c76347032d01ac37fbc587d0540ee6b0cf9b419d)."_

By now, we already have **four** things which are information not existing in CQL schema that we store in internal tags:

* `ProvisionedThroughput` parameters saved in `RCU_TAG_KEY`, `WCU_TAG_KEY`,
* TTL configuration of the table saved in  `TTL_TAG_KEY`
* Saving table creation time (information not available in CQL) in `TABLE_CREATION_TIME_TAG_KEY `

so it makes a lot of sense to add a fifth one. There's just a question what to call it and exactly what to write in it.

**Comment by radoslawcybulski** [comment]
@nyh i would write either bool flag or added column name (the latter so we didn't have to look it up ourselves).

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) i would write either bool flag or added column name (the latter so we didn't have to look it up ourselves).

You should think whether a bool called something like "range_key_is_artificial" (choose a better name ;_) is enough, or we might ever need something more general like a list of key columns to add or remove or I don't know what. But will we ever need anything more general? I can't think why.

**Comment by radoslawcybulski** [comment]
I went with ARTIFICIAL_HASH_KEY_ADDED_TO_GSI and ARTIFICIAL_RANGE_KEY_ADDED_TO_GSI. I can merge them into one, not sure if it's worth it.

**Comment by ScyllaPiotr** [comment]
To make it more connected with the problem / localized, maybe sth like `SKIP_FOR_DESCRIBE_TABLE` ?

**Comment by radoslawcybulski** [comment]
@ScyllaPiotr I think `what it is` is more useful than `why it's here`.

@nyh i went with storing column name inside a tag's key for simplicity and two tags for hash and range keys respectively.

Patch is here: https://github.com/scylladb/scylladb/pull/25978

**Comment by ScyllaPiotr** [comment]
> [@ScyllaPiotr](https://github.com/ScyllaPiotr) I think `what it is` is more useful than `why it's here`.

What I meant is that the word _artificial_ can have many meanings and implications. A 1st-time reader would have to grep for all the occurrences of `ARTIFICIAL_*KEY`, which is not infeasible, but what I wanted to achieve is to communicate better what this artificiality means already with the name. My proposal is not meant to clarify _why it's here_ (the reason), but to describe the effect of being here, the implications.

**Comment by nyh** [comment]
> I went with ARTIFICIAL_HASH_KEY_ADDED_TO_GSI and ARTIFICIAL_RANGE_KEY_ADDED_TO_GSI. I can merge them into one, not sure if it's worth it.

I don't think "arificial hash key" is possible - you can't have a table without a real hash key.

Please don't forget to consider additional scenarios beyond what I wrote in that single reproducer. In particular, I think you should consider, and test, the following three cases:

1. The base table has both hash and range keys (p, c). The GSI has just a hash key - x
2. The base table has both hash and range keys (p, c). The GSI has different hash and range keys - x1, x2 (this two-new-key-columns is an esoteric Alternator-only case not supported in CQL but we need to support it because it's supported on DynamoDB).
3. The base table has both hash and range keys (p,c). An LSI (not GSI) has a key (p, x).

Note that if I understand correctly, in case 1 and 2, we have two range keys that are artificial and shouldn't be displayed (p and c). case 3 has one (c).

There's one thing that is important to remember: DescribeTable has access not only to the GSI, but also to the base table, so it can also use the schema of the base table to improve its guesses without needing more information in the tag. For example, in case 2 we see three clustering columns - x2, p and c, but we know from the base table's key that p and c were also keys in the base, so since we have one which isn't (x2) and there can only be one range key, then x2 must be the only "real" range key and p and c are not and shouldn't be printed - and we ddn't need any tag to know that.

The case which I tested already is actually the most ambigous case: You have a base table with hash key p, and gsi with hash key x and range key p. You don't know whether range key p was requested by the user or added artificially - because both are actually allowed. This is why we need that boolean flag, and probably just one boolean flag is enough for all cases.

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) i went with storing column name inside a tag's key for simplicity and two tags for hash and range keys respectively.

Please see the comment I wrote above where I outline three more tests I think you should do, and why I think one (not two) boolean flags should be enough for everything. If you want to go with a name of an attribute - it's not really "simple" - it actually complicates some cases which have two, not one, artificial range keys - but in those cases the code can guess that these are artificial range keys and doesn't need extra flags (see my comment).

**Comment by radoslawcybulski** [comment]
> > I went with ARTIFICIAL_HASH_KEY_ADDED_TO_GSI and ARTIFICIAL_RANGE_KEY_ADDED_TO_GSI. I can merge them into one, not sure if it's worth it.
> 
> I don't think "arificial hash key" is possible - you can't have a table without a real hash key.

My enligh, oh my english. I meant that key is added by the "backend", not by the user. There's probably a better word for this.

> Please don't forget to consider additional scenarios beyond what I wrote in that single reproducer. In particular, I think you should consider, and test, the following three cases:
> 
>     1. The base table has both hash and range keys (p, c). The GSI has just a hash key - x
> 
>     2. The base table has both hash and range keys (p, c). The GSI has different hash and range keys - x1, x2 (this two-new-key-columns is an esoteric Alternator-only case not supported in CQL but we need to support it because it's supported on DynamoDB).
> 
>     3. The base table has both hash and range keys (p,c). An LSI (not GSI) has a key (p, x).

Will add some tests.

> There's one thing that is important to remember: DescribeTable has access not only to the GSI, but also to the base table, so it can also use the schema of the base table to improve its guesses without needing more information in the tag. For example, in case 2 we see three clustering columns - x2, p and c, but we know from the base table's key that p and c were also keys in the base, so since we have one which isn't (x2) and there can only be one range key, then x2 must be the only "real" range key and p and c are not and shouldn't be printed - and we ddn't need any tag to know that.

if i understand correctly, user might choose to specify `c` and `x2` as range or only `c` or only `x2`. Am i correct?

**Comment by radoslawcybulski** [comment]
> What I meant is that the word _artificial_ can have many meanings and implications. A 1st-time reader would have to grep for all the occurrences of `ARTIFICIAL_*KEY`, which is not infeasible, but what I wanted to achieve is to communicate better what this artificiality means already with the name. My proposal is not meant to clarify _why it's here_ (the reason), but to describe the effect of being here, the implications.

Hmm, interesting. I'd grep for exact value of one of keys (so for example `ARTIFICIAL_HASH_KEY_ADDED_TO_GSI`) and went from there.
I feel `SKIP_FOR_DESCRIBE_TABLE` identifier tells, `why it's there` (to notify others about keys that should be skip for describe table), while i'd rather have `what it is` type of information (additional key, added by the backend, not specified by user), from which it should follow, why we don't emit that key in describe table handling code.
Will try to figure out a better word, than artificial.

**Comment by nyh** [comment]
I would use a boolean SPURIOUS_RANGE_KEY, a boolean tag on the GSI table (not on the base table).  It would be only needed and, set, in one specific case: when the user did NOT ask for a range key on the gsi at all, but we were forced to add one (or two!) because of the base table's key columns.
In other case, namely when the user did choose a range key for the gsi/lsi (lsi always needs a range key!), this flag isn't needed - we already know that if there is more than one clustering key columns ib the view, the real  gsi/lsi is the only one which isn't a base key column. This logic is already in the code.

**Comment by radoslawcybulski** [comment]
@nyh The logic you posted is complicated and error prone, hard to follow and i literally will have to spend next several minutes to verify you're right. Do we need this? My solution is straightforward, everyone understands, what it does. Do we have additional penalty for tags we're using? If so, maybe we should improve tags handling instead?

**Comment by nyh** [comment]
The logic I mentioned is already 90% in the code. Your logic is different, so you'll need to test it in different cases, not just the specific one that the current logic fails on. This is why I asked you to write more tests (and mentioned which ones I think are important) that verify the changes don't regress what is already working.

**Comment by radoslawcybulski** [comment]
Where is the logic you're talking about? I only found this:

```
            if  (hash_key != view_hash_key && hash_key != view_range_key) {
                add_column(view_builder, hash_key, *attribute_definitions, column_kind::clustering_key);
                tags[ARTIFICIAL_HASH_KEY_ADDED_TO_GSI_TAG_KEY] = hash_key;
            }
            if  (!range_key.empty() && range_key != view_hash_key && range_key != view_range_key) {
                add_column(view_builder, range_key, *attribute_definitions, column_kind::clustering_key);
                tags[ARTIFICIAL_RANGE_KEY_ADDED_TO_GSI_TAG_KEY] = range_key;
            }
```

Should i rely on the implementation detail that user's key, if present, is always first?

**Comment by nyh** [comment]
This is indeed the current logic in the code: we first add the user's requested range key (if any) and then the other "artificial" range keys (the base key columns not part of the gsi key).

**Comment by nyh** [comment]
Let me try to explain the existing logic (which, as I explained in my review of your patch https://github.com/scylladb/scylladb/pull/25978, you may need to continue supporting to allow the new code to DescribeTable old tables):

We know two facts:
1. In DynamoDB, each table or index (GSI/LSI) has one or two key components: One hash key, and one or none range key (for LSI, the "none" case is not possible). You can't have more than one range key columns (which CQL does allow).
2. In Scylla's Materialized Views, each one of the base's hash/range keys must be present as some key column (of any type) in each view. So for example, if the has has key columns p,c and you witn a view with key columns p,x, then c must be added - as another clustering key column - in this case you'll have two clustering columns x,c.

Because of fact 2, we may need to add one or two "artificial" clustering keys to the MV implementing the GSI/LSI.
If the GSI has a clustering key (aka range key) - note that LSI always must have one - then it's easy to know which is the real GSI range key and which are the added-just-for-MV - the first one is the real key, and any following ones are the fake ones, and shouldn't be printed. This is why the "break" in the existing implementation (the break with a FIXME in front of it) exists.

This is the existing logic, and it only breaks when the GSI doesn't have a range key at all and only has a hash key. In that case, when you see a clustering key in a materialized view, you have no idea if it's a real one or a fake one.

So to fix the existing implementation without changing it completely (and risking breaking compatibility with already-existing tables), you only need to add a tag that says "this GSI does not have a clustering key, and the ones we added weren't requested by the user". In all other cases, you don't need any tag at all - the existing logic already handles everything.

**Comment by nyh** [comment]
> Should i rely on the implementation detail that user's key, if present, is always first?

This is what the current implementation guarantees, and assumes, indeed.

I think an alternative to this assumption is to check which of the range keys aren't one of the base key columns - if we find such a range key we know it was a real range key chosen by the user.
But since we know the order of the range keys, I it think it's easier just to make use of it and don't work harder.

**Comment by ScyllaPiotr** [comment]
> 1. In DynamoDB, each table or index (GSI/LSI) has one or two key components: One hash key, and one or none range key. You can't have more than one range key columns (which CQL does allow).

@nyh When I read this the 1st time I understood that LSI can have zero range (or ordering) keys, which is not true. You even correct this in the same message. Could you correct your post in the cited place too, so it's less ambiguous?

> But since we know the order of the range keys, I it think it's easier just to make use of it and don't work harder.

@nyh is this order documented, at least in the source? Since Radek is going to start using this rule, best to write it down somewhere.

**Comment by nyh** [comment]
> > 1. In DynamoDB, each table or index (GSI/LSI) has one or two key components: One hash key, and one or none range key. You can't have more than one range key columns (which CQL does allow).
> 
> [@nyh](https://github.com/nyh) When I read this the 1st time I understood that LSI can have zero range (or ordering) keys, which is not true. You even correct this in the same message. Could you correct your post in the cited place too, so it's less ambiguous?

Which cited place - in my comment which you cited? Ok, I'll edit it.

> 
> > But since we know the order of the range keys, I it think it's easier just to make use of it and don't work harder.
> 
> [@nyh](https://github.com/nyh) is this order documented, at least in the source? Since Radek is going to start using this rule, best to write it down somewhere.

We do not have official documentation on how Alternator's "data model", i.e., how Alternator tables translate to storage on underlying Scylla tables. In the original Alternator design document (https://docs.google.com/document/d/1i4yjF5OSAazAY_-T8CBce9-2ykW4twx_E_Nt2zDoOVs/edit?tab=t.0) I did not anticipate this issue, and did not discuss it.

It's important to understand that as far as the Alternator data model is concerned, the order of clustering keys is just a tiny part of this - there's also column names (like ":attrs"), encoding of individual cells, tags, how GSIs and LSIs keys work, and a bunch of other things - regardless of whether any of that is documented in any official document, all of it is very important for backward compatibility with pre-existing tables created in previous releases. It doesn't mean we can't change anything (we just changed how LSI and GSI keys are stored recently!), but if we do, we need to think carefully if the change is backward compatible. In the "worst" case we can always have two code paths - for some "old way" and for the "new way", but in the present case I simply don't think we need to do this - we don't really need to change the old way at all, we just need to add an override for the one case where the old way was broken (and of course, in the specific case it was broken it is fine if it remains broken in the same way after the upgrade - as long as new tables no longer have this bug).

**Comment by nyh** [comment]
Existing issue on the need to document Alternator's "on disk format": https://github.com/scylladb/scylladb/issues/19770

**Comment by ScyllaPiotr** [comment]
> It's important to understand that as far as the Alternator data model is concerned, the order of clustering keys is just a tiny part of this

True, asking to document every tiny implementation decision would be ridiculous. What I wanted to suggest is for this new dependency (that the mechanism for DescribeTable to report a GSI range key or not relies on the assumption that the user-requested range key of the GSI, if present, is always the 1st one in the underlying MV) to become now documented in the source.
On 2nd thought, maybe instead @radoslawcybulski could put a comment that he actually utilized this rule in his new code.

**Comment by nyh** [comment]
> > It's important to understand that as far as the Alternator data model is concerned, the order of clustering keys is just a tiny part of this
> 
> True, asking to document every tiny implementation decision would be ridiculous.

It's not ridiculous, it's actually important. More than a year ago I opened https://github.com/scylladb/scylladb/issues/19770 about it and Yaniv marked it P1. But as usual, since everything is urgent and important and what's "most urgent" and "most important" changes every week, it never got done.

As I noted, these details were supposed to be in the original Alternator design document, but as meticulous as I am - I wasn't **that** meticulous and I missed discussing the issue of those "extra clustering keys" needed by Scylla MV - and it was only discovered during the implementation. And beyond that - there are things that changed after the original design so not everything in the original design document is still up-to-date.

> What I wanted to suggest is for this new dependency (that the mechanism for DescribeTable to report a GSI range key or not relies on the assumption that the user-requested range key of the GSI, if present, is always the 1st one in the underlying MV) to become now documented in the source.

My point is that it's not a new dependency, it was present in the existing code already.
And definitely - I did have a comment on the "break" that implements the DescribeTable part of this, but it needs to be rewritten to explain the real situation, and another comment might be added to the place where we build the view. In my review, I asked @radoslawcybulski to not remove this "break" for backward-compatibility purposes (and also sanity reasons - DynamoDB doesn't allow multiple clustering keys so DescribeTable should list more than one!), and to rewrite the comment in front of it.

> On 2nd thought, maybe instead [@radoslawcybulski](https://github.com/radoslawcybulski) could put a comment that he actually utilized this rule in his new code.
